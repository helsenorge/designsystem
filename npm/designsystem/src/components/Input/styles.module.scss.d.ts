export type Styles = {
  'content-wrapper': string;
  'content-wrapper__input': string;
  'content-wrapper__input--bigform': string;
  'content-wrapper__input--disabled': string;
  'content-wrapper--bigform': string;
  'content-wrapper--disabled': string;
  'content-wrapper--invalid': string;
  'content-wrapper--on-blueberry': string;
  'content-wrapper--on-dark': string;
  'content-wrapper--transparent': string;
  'content-wrapper--with-icon': string;
  'input-wrapper': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
