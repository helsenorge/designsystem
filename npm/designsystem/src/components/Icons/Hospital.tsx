import React from 'react';
import { getIcon, SvgPathProps } from './Icon';

const Hospital: React.FC<SvgPathProps> = ({ size, isHovered }: SvgPathProps): JSX.Element => {
  const normal = (
    <path
      fillRule={'evenodd'}
      d="M13.554 28.445h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zm1 6.312h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zm5.986-16.908h1v-3.158h3.159v-1H19.54v-3.159h-1v3.159h-3.159v1h3.159v3.158zM30.791 39.72h8.37V20.682h-8.37V39.72zm-22.202 0h20.903V10.547H8.589V39.72zm22.203-20.338V9.247H7.289V41.02h33.173V19.382h-9.67zm-9.139 9.063h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zm1 6.312h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zM10.869 7.702h16.343v-1.3H10.869v1.3z"
    />
  );

  const normalHover = (
    <path
      fillRule={'evenodd'}
      d="M13.554 28.445h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zm1 6.312h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zm5.986-15.908h1v-4.158h4.159v-1H19.54v-4.159h-1v4.159h-4.159v1h4.159v4.158zM30.791 39.72h8.37V20.682h-8.37V39.72zm-22.202 0h20.903V10.547H8.589V39.72zm22.203-20.338V9.247H7.289V41.02h33.173V19.382h-9.67zm-9.139 9.063h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zm1 6.312h2.873v-3.243h-2.873v3.243zm-1 1h4.873v-5.243h-4.873v5.243zM10.869 7.702h16.343v-1.3H10.869v1.3z"
    />
  );

  const small = (
    <path
      fillRule={'evenodd'}
      d="M21.527 35.62h2.862v-3.233h-2.862v3.234zm-1.01 1.01H25.4v-5.254h-4.884v5.254zm-7.09-1.01h2.863v-3.233h-2.863v3.234zm-1.01 1.01H17.3v-5.254h-4.883v5.254zm9.11-8.322h2.862v-3.232h-2.862v3.232zm-1.01 1.01H25.4v-5.254h-4.884v5.254zm-7.09-1.01h2.863v-3.232h-2.863v3.232zm-1.01 1.01H17.3v-5.253h-4.883v5.254zm5.86-9.6h1.264V16.69h3.027v-1.263H19.54V12.4h-1.263v3.027H15.25v1.263h3.028v3.028zm12.365 19.888h8.406V20.53h-8.406v19.075zm-22.204 0h20.94V9.132H8.439v30.474zm22.204-20.338V7.87H7.175v33H40.31V19.267h-9.67z"
    />
  );

  const smallHover = (
    <path
      fillRule={'evenodd'}
      d="M22.158 35.62h2.863v-3.233h-2.863v3.234zm-1.01 1.01h4.883v-5.254h-4.883v5.254zm-7.09-1.01h2.863v-3.233H14.06v3.234zm-1.01 1.01h4.884v-5.254h-4.884v5.254zm9.11-8.322h2.863v-3.232h-2.863v3.232zm-1.01 1.01h4.883v-5.254h-4.883v5.254zm-7.09-1.01h2.863v-3.232H14.06v3.232zm-1.01 1.01h4.884v-5.253h-4.884v5.254zm5.861-9.6h1.264V16.69h3.026v-1.263h-3.026V12.4h-1.264v3.027h-3.027v1.263h3.027v3.028zm12.364 19.888h8.407V20.53h-8.407v19.075zm-22.204 0H30.01V9.132H9.07v30.474zm22.204-20.338V7.87H7.806v33h33.137V19.267h-9.67z"
    />
  );

  return getIcon(size, isHovered, normal, normalHover, small, smallHover);
};

export default Hospital;
