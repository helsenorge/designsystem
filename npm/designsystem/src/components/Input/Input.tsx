import React, { useRef } from 'react';
import cn from 'classnames';

import { FormMode, FormVariant, AnalyticsId } from '../../constants';
import { uuid } from '../../utils/uuid';
import Icon, { IconSize, SvgIcon } from '../Icons';
import { getColor } from '../../theme/currys';
import { Breakpoint, useBreakpoint } from '../../hooks/useBreakpoint';

import ErrorWrapper from '../ErrorWrapper';

import styles from './styles.module.scss';

export interface InputProps
  extends Pick<
    React.InputHTMLAttributes<HTMLInputElement>,
    'disabled' | 'readOnly' | 'autoComplete' | 'name' | 'placeholder' | 'defaultValue' | 'required'
  > {
  /**  HMTL Input type */
  type?: keyof typeof InputTypes;
  /** input id */
  inputId?: string;
  /** Width of input field in characters (approximate) */
  width?: number;
  /** If true, the component will be transparent. */
  transparent?: boolean;
  /** Icon to be displayed next to the input field */
  icon?: SvgIcon;
  /** Places the icon to the right */
  iconRight?: boolean;
  /** Changes the color profile of the input */
  mode?: keyof typeof FormMode;
  /** Changes the visuals of the input */
  variant?: keyof typeof FormVariant;
  /** Label of the input */
  label?: string;
  /** Activates Error style for the input */
  error?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Component shown after label */
  afterLabelChildren?: React.ReactNode;
  /** Component shown under label */
  belowLabelChildren?: React.ReactNode;
}

export enum InputTypes {
  text = 'text',
  number = 'number',
  email = 'email',
  password = 'password',
  search = 'search',
  tel = 'tel',
  url = 'url',
}

export const AVERAGE_CHARACTER_WIDTH_PX = 12;

const getInputMaxWidth = (characters: number, hasIcon: boolean, iconSize: number): string => {
  const paddingWidth = hasIcon ? '1.5rem' : '2rem';
  const iconWidth = hasIcon ? `${iconSize}px` : '0px';
  const borderWidth = '4px';

  return `calc(${characters * AVERAGE_CHARACTER_WIDTH_PX}px + ${paddingWidth} + ${iconWidth} + ${borderWidth})`;
};

const Input = React.forwardRef((props: InputProps, ref: React.Ref<HTMLInputElement>) => {
  const {
    defaultValue,
    placeholder,
    type = InputTypes.text,
    inputId = uuid(),
    name,
    transparent = false,
    icon,
    iconRight,
    mode,
    variant,
    label,
    error,
    errorText,
    testId,
    disabled,
    readOnly,
    autoComplete,
    afterLabelChildren,
    belowLabelChildren,
    width,
    required,
  } = props;
  const breakpoint = useBreakpoint();
  const contentWrapperRef = useRef<HTMLDivElement>(null);

  const onDark = mode === FormMode.ondark;
  const onBlueberry = mode === FormMode.onblueberry;
  const onError = mode === FormMode.oninvalid || errorText || error;
  const bigForm = variant === FormVariant.bigform;
  const isTransparent = transparent && mode !== FormMode.ondark && !onError;

  const inputWrapperClass = cn(styles['input-wrapper']);

  const labelWrapperClass = cn(styles['input-wrapper__label-wrapper'], {
    [styles['input-wrapper__label-wrapper--on-dark']]: onDark,
    [styles['input-wrapper__label-wrapper--disabled']]: disabled,
  });

  const contentWrapperClass = cn(styles['content-wrapper'], {
    [styles['content-wrapper--transparent']]: isTransparent,
    [styles['content-wrapper--on-blueberry']]: onBlueberry,
    [styles['content-wrapper--on-dark']]: onDark,
    [styles['content-wrapper--invalid']]: onError,
    [styles['content-wrapper--bigform']]: bigForm,
    [styles['content-wrapper--disabled']]: disabled,
    [styles['content-wrapper--with-icon']]: icon,
  });

  const inputClass = cn(styles['content-wrapper__input'], {
    [styles['content-wrapper__input--bigform']]: bigForm,
    [styles['content-wrapper__input--disabled']]: disabled,
  });

  const iconColor = disabled ? getColor('neutral', 500) : getColor('black');
  const iconSize = breakpoint === Breakpoint.xs || !bigForm ? IconSize.XSmall : IconSize.Small;

  const renderIcon = () => {
    return icon !== undefined ? <Icon color={iconColor} size={iconSize} svgIcon={icon} /> : null;
  };

  const handleClick = (): void => {
    if (contentWrapperRef && contentWrapperRef.current && icon) {
      const selectedChild = iconRight ? 0 : 1;
      const input = contentWrapperRef.current.children[selectedChild] as HTMLInputElement;
      input.focus();
    }
  };

  const maxWidth = width ? getInputMaxWidth(width, !!icon, iconSize) : undefined;

  return (
    <ErrorWrapper errorText={errorText}>
      <div data-testid={testId} data-analyticsid={AnalyticsId.Input} className={inputWrapperClass}>
        {label && (
          <div className={labelWrapperClass}>
            <label htmlFor={inputId}>{label}</label>
            {afterLabelChildren && <div className={styles['input-wrapper__after-label-children']}>{afterLabelChildren}</div>}
          </div>
        )}
        {belowLabelChildren && <div>{belowLabelChildren}</div>}
        <div onClick={handleClick} ref={contentWrapperRef} className={contentWrapperClass} style={{ maxWidth }}>
          {!iconRight && renderIcon()}
          <input
            name={name}
            type={type}
            defaultValue={defaultValue}
            id={inputId}
            className={inputClass}
            ref={ref}
            aria-invalid={!!onError}
            disabled={disabled}
            placeholder={placeholder}
            readOnly={readOnly}
            autoComplete={autoComplete || 'off'}
            required={required}
          />
          {iconRight && renderIcon()}
        </div>
      </div>
    </ErrorWrapper>
  );
});

export default Input;
