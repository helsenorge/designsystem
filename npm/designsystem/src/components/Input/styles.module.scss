@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';
@import '../../scss/input';

.input-wrapper {
  @include input-wrapper;

  margin: getSpacer(l) 0;
}

.input-wrapper:first-of-type {
  margin-top: 0;
}

.input-wrapper:last-of-type {
  margin-bottom: 0;
}

.content-wrapper {
  display: flex;
  position: relative;
  width: fit-content;
}

.input-container {
  @include input-container;

  padding: 0.5rem 0;
  align-items: center;

  &--with-icon {
    padding: 1px 0;

    @media (min-width: map.get($grid-breakpoints, md)) {
      padding: 0.375rem 0;
    }
  }

  &--with-icon#{&}--bigform {
    padding: getSpacer(2xs) 0;
  }

  &__input {
    line-height: getSpacer(m);
    height: getSpacer(m);

    &--bigform {
      height: 2.5rem;
    }

    @media (min-width: map.get($grid-breakpoints, md)) {
      line-height: 1.625rem;
      height: 2.125rem;
      font-size: $font-size-md;

      &--bigform {
        line-height: 1.75rem;
        height: 3.5rem;
      }
    }

    &__icon {
      cursor: pointer;
    }
  }
}
