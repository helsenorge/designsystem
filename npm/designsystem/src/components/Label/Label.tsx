import React, { FunctionComponent } from 'react';

import cn from 'classnames';

import { Sublabel, SublabelProps } from './SubLabel';
import { AnalyticsId, FormOnColor } from '../../constants';
import { isComponent } from '../../utils/component';
import FormFieldTag, { FormFieldTagLevel } from '../FormFieldTag';
import Spacer from '../Spacer';
import StatusDot, { StatusDotProps } from '../StatusDot';

import styles from './styles.module.scss';

export type LabelText = {
  hideFromScreenReader?: boolean;
  text: string;
  type?: 'normal' | 'subdued';
};

export type LabelTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'label' | 'p';

export interface LabelProps {
  /** Component shown after label - discourage use of this */
  afterLabelChildren?: React.ReactNode;
  /** Sets the content of the Label */
  children?: React.ReactNode;
  /** Sets a tag that describes whether the form field is required or optional */
  formFieldLevel?: FormFieldTagLevel;
  /** Adds custom classes to the label tag. */
  labelClassName?: string;
  /** Adds custom classes to the label text. */
  labelTextClassName?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Id that is put on the "for" attribute of the label */
  htmlFor?: string;
  /** Changes the underlying element of the label */
  htmlMarkup?: LabelTags;
  /** Id som plasseres p√• <label/> */
  labelId?: string;
  /** Array of main label strings. Can be of type semibold or normal */
  labelTexts?: LabelText[];
  /** Array of sublabel strings. Can be of type semibold or normal */
  onColor?: keyof typeof FormOnColor;
  /** StatusDot placed underneath the last sublabel */
  statusDot?: React.ReactNode;
  /** Sublabel component */
  sublabel?: React.ReactNode;
  /** Adds custom classes to the div wrapping the sublabels. */
  sublabelWrapperClassName?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const getLabelText = (label: React.ReactNode): string => {
  let allLabelText = '';

  if (isComponent<LabelProps>(label, Label)) {
    label.props.labelTexts?.forEach(labelText => {
      allLabelText += !labelText.hideFromScreenReader ? labelText.text : '';
    });
  }

  return allLabelText;
};

export const renderLabel = (label: React.ReactNode, inputId: string, onColor: FormOnColor, markup?: LabelTags): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            htmlMarkup: markup || 'label',
            onColor,
          })
        : typeof label === 'string' && <Label labelTexts={[{ text: label, type: 'normal' }]} htmlFor={inputId} onColor={onColor} />}
    </>
  );
};

export const renderLabelAsParent = (
  label: React.ReactNode,
  children: React.ReactNode,
  inputId: string,
  onColor: FormOnColor,
  labelClassName?: string,
  labelTextClassName?: string,
  sublabelWrapperClassName?: string,
  large?: boolean,
  markup?: LabelTags
): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            onColor,
            children: children,
            labelClassName: cn(labelClassName, label.props.labelClassName),
            labelTextClassName: labelTextClassName,
            htmlMarkup: markup || 'label',
            sublabelWrapperClassName: sublabelWrapperClassName,
            sublabel: large ? undefined : label.props.sublabel,
            statusDot: large ? undefined : label.props.statusDot,
          })
        : typeof label === 'string' && (
            <Label
              labelTexts={[{ text: label, type: 'subdued' }]}
              htmlFor={inputId}
              onColor={onColor}
              htmlMarkup={markup || 'label'}
              labelClassName={labelClassName}
              labelTextClassName={labelTextClassName}
              sublabelWrapperClassName={sublabelWrapperClassName}
            >
              {children}
            </Label>
          )}
    </>
  );
};

const Label: FunctionComponent<LabelProps> = ({
  afterLabelChildren,
  children,
  className,
  formFieldLevel,
  htmlFor,
  htmlMarkup = 'label',
  labelClassName,
  labelTextClassName,
  labelId,
  labelTexts,
  onColor = FormOnColor.onwhite,
  statusDot,
  sublabel,
  sublabelWrapperClassName,
  testId,
}) => {
  const hasChildren = children && typeof children !== 'undefined';
  const labelWrapperClasses = cn(
    styles['label-wrapper'],
    { [styles['label-wrapper--no-bottom-margin']]: hasChildren, [styles['label-wrapper--after-label-children']]: afterLabelChildren },
    className
  );

  const mapLabels = (): React.ReactNode => {
    if (typeof labelTexts === 'undefined') return null;

    return labelTexts.map((labelText, index) => {
      const labelClasses = cn(
        styles.label,
        {
          [styles['label--subdued']]: labelText.type === 'subdued',
          [styles['label--on-dark']]: onColor === FormOnColor.ondark,
        },
        labelTextClassName
      );
      return (
        <span aria-hidden={labelText.hideFromScreenReader} className={labelClasses} key={index}>
          {labelText.text}
        </span>
      );
    });
  };
  const CustomTag = htmlMarkup;

  return (
    <div className={labelWrapperClasses}>
      <div>
        <CustomTag className={labelClassName} id={labelId} htmlFor={htmlFor} data-testid={testId} data-analyticsid={AnalyticsId.Label}>
          <span className={styles['label-content-wrapper']}>
            {children}
            <span className={styles.label__texts}>{mapLabels()}</span>
          </span>
        </CustomTag>
        <div className={sublabelWrapperClassName}>
          {formFieldLevel && <FormFieldTag level={formFieldLevel} />}
          {sublabel &&
            isComponent<SublabelProps>(sublabel, Sublabel) &&
            React.cloneElement(sublabel, {
              onColor: onColor as FormOnColor,
            })}
          {statusDot && isComponent<StatusDotProps>(statusDot, StatusDot) && (
            <>
              <Spacer size={'3xs'} />
              {React.cloneElement(statusDot, {
                onColor: onColor === FormOnColor.ondark ? 'ondark' : 'onwhite',
              })}
            </>
          )}
        </div>
      </div>
      {afterLabelChildren && <div className={styles['after-label-children']}>{afterLabelChildren}</div>}
    </div>
  );
};

export default Label;
