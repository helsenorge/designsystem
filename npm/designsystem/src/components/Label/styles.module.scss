@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';

.label-wrapper {
  margin-bottom: getSpacer(2xs);
  line-height: $lineheight-size-xs;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    line-height: $lineheight-size-sm;
  }

  &--no-bottom-margin {
    margin-bottom: 0;
  }

  &--after-label-children {
    display: flex;
    align-items: flex-end;
  }
}

.label-content-wrapper {
  display: flex;
}

.label {
  font-size: $font-size-sm;
  font-weight: 400;
  line-height: $lineheight-size-xs;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    font-size: $font-size-md;
    line-height: $lineheight-size-sm;
  }

  &--semibold {
    font-weight: 600;
  }

  &--sublabel {
    display: block;
    font-size: $font-size-xs;
    line-height: 1.25rem;
    color: $neutral800;

    @media (min-width: map.get($grid-breakpoints, sm)) {
      font-size: $font-size-sm;
      line-height: $lineheight-size-xs;
    }
  }

  &--on-dark {
    color: $white;
  }

  &::after {
    content: ' ';
  }
}

.after-label-children {
  padding-left: getSpacer(xs);
}
