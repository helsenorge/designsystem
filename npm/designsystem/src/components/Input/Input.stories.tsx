import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Input, { InputProps, InputTypes } from './Input';
import { FormMode, FormSize } from '../../constants';
import Docs from '../../docs';
import Hospital from '../Icons/Hospital';
import Label from '../Label/Label';

type InputWithAndCustomArgs = React.ComponentProps<typeof Input> & {
  showIcon: boolean;
};

const meta = {
  title: '@helsenorge/designsystem-react/Components/Input',
  component: Input,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={Input} />,
      description: {
        component:
          'Som innbygger vil jeg kunne skrive inn korte tekster slik at jeg kan gjøre mine oppgaver.<br><br>Bruksområde: For å la innbygger skrive inn en tekst (enkeltlinje, begrenset mengde)',
      },
    },
  },
  args: {
    width: undefined,
    size: FormSize.medium,
    transparent: false,
    type: InputTypes.text,
    disabled: false,
    mode: FormMode.onwhite,
    showIcon: true,
    iconRight: false,
    readOnly: false,
    autoComplete: '',
    name: 'input',
    required: false,
    maxText: 'tegn',
  },
  argTypes: {
    defaultValue: {
      control: 'text',
    },
    placeholder: {
      control: 'text',
    },
    width: {
      control: 'number',
    },
    size: {
      control: 'select',
      options: FormSize,
    },
    transparent: {
      control: 'boolean',
    },
    type: {
      control: 'select',
      options: InputTypes,
    },
    disabled: {
      control: 'boolean',
    },
    mode: {
      control: 'select',
      options: FormMode,
    },
    showIcon: {
      control: 'boolean',
    },
    iconRight: {
      control: 'boolean',
    },
    readOnly: {
      control: 'boolean',
    },
    autoComplete: {
      control: 'text',
    },
    name: {
      control: 'text',
    },
    required: {
      control: 'boolean',
    },
    maxCharacters: {
      control: 'number',
    },
    maxText: {
      control: 'text',
    },
  },
} satisfies Meta<InputWithAndCustomArgs>;

export default meta;

type Story = StoryObj<InputWithAndCustomArgs>;

export const Default: Story = {
  render: ({ showIcon, ...rest }) => {
    const inputId = 'input-testid';

    return (
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} htmlFor={inputId} />}
        inputId={inputId}
        icon={showIcon ? Hospital : undefined}
      />
    );
  },
};

export const MultipleExamples: Story = {
  render: ({ showIcon, ...rest }) => (
    <>
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        icon={showIcon ? Hospital : undefined}
      />
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        icon={showIcon ? Hospital : undefined}
      />
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        icon={showIcon ? Hospital : undefined}
      />
    </>
  ),
};

export const MaxCharacters: Story = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  render: ({ showIcon, ...rest }: InputProps & { showIcon?: boolean }) => (
    <>
      <Input {...rest} label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />} maxCharacters={10} width={10} />
      <Input {...rest} label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />} maxCharacters={50} width={50} />
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        maxCharacters={50}
        width={50}
        defaultValue="test"
      />
    </>
  ),
};

export const AfterInputChildren: Story = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  render: ({ showIcon, ...rest }: InputProps & { showIcon?: boolean }) => (
    <>
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        afterInputChildren={<div style={{ marginTop: '1rem' }}>{'*Ikke oppgi personsensitiv informasjon'}</div>}
      />
      <Input
        {...rest}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} />}
        afterInputChildren={'Tekst uten innsendt div'}
      />
    </>
  ),
};

export const BaseIncrementValue: Story = {
  render: ({ showIcon, ...rest }: InputProps & { showIcon?: boolean }) => {
    const inputId = 'input-testid';

    return (
      <Input
        {...rest}
        type={'number'}
        baseIncrementValue={1990}
        label={<Label labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]} htmlFor={inputId} />}
        inputId={inputId}
        icon={showIcon ? Hospital : undefined}
      />
    );
  },
};
