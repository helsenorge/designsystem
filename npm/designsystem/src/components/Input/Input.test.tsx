import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Input, { InputTypes } from './Input';
import { FormMode, FormVariant } from '../../constants';
import Hospital from '../Icons/Hospital';
import Label from '../Label';

describe('Gitt at Input skal vises', (): void => {
  describe('Når Input rendres', (): void => {
    test('Så vises Input', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} />);

      const label = screen.getByText('Skriv noe');
      expect(label).toBeVisible();

      const input = screen.getByRole('textbox');
      expect(input).toBeVisible();
      expect(input).toHaveClass('content-wrapper__input');
    });
  });

  describe('Når disabled er true', (): void => {
    test('Så vises Input som disabled', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe!' }]} />} disabled />);

      const input = screen.getByRole('textbox');
      expect(input).toBeDisabled();
    });
  });

  describe('Når FormMode er onblueberry', (): void => {
    test('Så vises Input med onblueberry styling', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} mode={FormMode.onblueberry} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--on-blueberry');
    });
  });
  describe('Når mode er ondark', (): void => {
    test('Så vises Input med ondark styling', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} mode={FormMode.ondark} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--on-dark');
    });
  });
  describe('Når transparent er true', (): void => {
    test('Så vises Input med transparent styling', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} transparent />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--transparent');
    });
  });
  describe('Når formvariant er bigform', (): void => {
    test('Så vises Input med bigform styling', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} variant={FormVariant.bigform} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--bigform');
    });
  });

  describe('Når Input har en default value', (): void => {
    test('Så skal default value vises', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} defaultValue={'default text'} />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveValue('default text');
    });
  });

  describe('Når Input har en placeholder value', (): void => {
    test('Så skal placeholder verdien vises', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} placeholder={'placeholder  text'} />);
      const input = screen.getByPlaceholderText('placeholder text');
      expect(input).toBeVisible();
    });
  });

  describe('Når type settes', (): void => {
    test('Så skal input ha den typen satt', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} type={InputTypes.number} />);
      const input = screen.getByRole('spinbutton');
      expect(input).toBeVisible();
    });
  });

  describe('Når icon settes', (): void => {
    test('Så rendres input med icon', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} icon={Hospital} />);
      const svg = screen.getByRole('textbox').previousSibling;
      expect(svg).toHaveClass('hnds-style-icon');
    });
  });

  describe('Når iconRight er true', (): void => {
    test('Så rendres input med icon til høyre', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} icon={Hospital} iconRight />);
      const svg = screen.getByRole('textbox').nextSibling;
      expect(svg).toHaveClass('hnds-style-icon');
    });
  });

  describe('Når Input har en error', (): void => {
    test('Så skal invalid styling brukes', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'Skriv noe' }]} />} error />);
      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--invalid');
    });
  });

  describe('Når disabled er satt', (): void => {
    test('Så er input disabled', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} disabled />);

      const input = screen.getByLabelText('En fin label');
      expect(input).toBeDisabled();
    });
  });

  describe('Når readOnly er satt', (): void => {
    test('Så er input readOnly', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} readOnly />);

      const input = screen.getByLabelText('En fin label');
      expect(input).toHaveAttribute('readonly', '');
    });
  });

  describe('Når autoComplete er satt til on', (): void => {
    test('Så har input autoComplete=on', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} autoComplete="on" />);

      const input = screen.getByLabelText('En fin label');
      expect(input).toHaveAttribute('autoComplete', 'on');
    });
  });

  describe('Når name-prop er satt', (): void => {
    test('Så har input riktig name', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} name="custom-name" />);

      const input = screen.getByLabelText('En fin label');
      expect(input).toHaveAttribute('name', 'custom-name');
    });
  });

  describe('Når placeholder-prop er satt', (): void => {
    test('Så har input riktig placeholder', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} placeholder="custom-placeholder" />);

      const input = screen.getByPlaceholderText('custom-placeholder');
      expect(input).toBeVisible();
    });
  });

  describe('Når defaultValue-prop er satt', (): void => {
    test('Så har input riktig value', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} defaultValue="custom-value" />);

      const input = screen.getByLabelText('En fin label');
      expect(input).toHaveValue('custom-value');
    });
  });

  describe('Når required er satt', (): void => {
    test('Så er input required', (): void => {
      render(<Input label={<Label labelTexts={[{ text: 'En fin label' }]} />} required />);

      const input = screen.getByLabelText('En fin label');
      expect(input).toBeRequired();
    });
  });

  describe('Når antall tegn skal vises', (): void => {
    describe('Når man skriver', (): void => {
      test('Så endres teksten og antall tegn', async (): Promise<void> => {
        render(<Input label={<Label labelTexts={[{ text: 'Skriv din historie her' }]} />} maxCharacters={50} />);

        const input = screen.getByLabelText('Skriv din historie her');

        expect(screen.getByText('0/50 tegn')).toBeVisible();

        await userEvent.type(input, 'Jeg tester teksten her.');

        expect(input).toHaveValue('Jeg tester teksten her.');
        expect(screen.getByText('23/50 tegn')).toBeVisible();
      });
    });

    describe('Når man skriver for mange tegn', (): void => {
      test('Så indikeres det at input er ugyldig', async (): Promise<void> => {
        render(<Input label={<Label labelTexts={[{ text: 'Skriv din historie her' }]} />} maxCharacters={10} />);

        const input = screen.getByLabelText('Skriv din historie her');

        expect(screen.getByText('0/10 tegn')).toBeVisible();

        await userEvent.type(input, 'Jeg tester teksten her.');

        expect(input).toHaveValue('Jeg tester teksten her.');
        expect(screen.getByText('23/10 tegn')).toBeVisible();
        expect(input).toHaveAttribute('aria-invalid', 'true');
      });
    });
  });
});
