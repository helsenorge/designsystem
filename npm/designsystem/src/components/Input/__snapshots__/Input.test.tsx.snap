// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Input skal vises Når Input rendres Så vises Input 1`] = `
<div>
  <div
    class="error-wrapper"
  >
    <div
      class="input-wrapper"
      data-analyticsid="input"
    >
      <div
        class="input-wrapper__label-wrapper"
      >
        <label
          for="test"
        >
          Skriv noe
        </label>
      </div>
      <div
        class="content-wrapper"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="content-wrapper__input"
          id="test"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
